### PR TITLE
Table primary columns in cyclic dependencies

### LINKED ISSUE #3291: [Bug] Table Column Settings Getting Reset
## Description

When the table data changes, the column settings are reset. We did this to avoid bloating the DSL with older column settings.
This has lead to many poor experiences of lost work. We should instead show columns that have configurations but are not in the table separately, maybe as hidden columns, and allow users to manually delete them
@Debsourabh please help here

### LINKED ISSUE #3440: [Bug]Losing primary columns custom column formatting after getting a cyclical dependency issue on table columns

### PR DESCRIPTION
**Pull request by riodeuno**
## Description
When there is a cyclic dependency, the table data is processed as a string, leading to incorrect column computations and a render loop. Fixed this by bailing out when table data is a string, as further computations are moot

Fixes #3440 
Fixes #3291 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Cypress test pending.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by hetunandu** [comment]
We should try to do this in the validation layer. Table should not get the string in the first place

**Comment by riodeuno** [comment]
@hetunandu PR updated.

File: app/client/src/workers/validations.ts [hunk 1]
```diff
@@ -247,7 +247,7 @@ export const VALIDATORS: Record<ValidationType, Validator> = {
       return {
         isValid,
         parsed: [],
-        transformed,
+        transformed: Array.isArray(transformed) ? transformed : [],
```
Comment (hetunandu) [inline app/client/src/workers/validations.ts:250]: Are you sure this is the correct code path? `transformed` values are only used in evaluated value pane. I think the problem is that is getting passed from this function as valid

File: app/client/src/widgets/TableWidget/index.tsx [hunk 2]
```diff
@@ -60,6 +60,7 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
       defaultSearchText: VALIDATION_TYPES.TEXT,
       defaultSelectedRow: VALIDATION_TYPES.DEFAULT_SELECTED_ROW,
       pageSize: VALIDATION_TYPES.NUMBER,
+      sanitizedTableData: VALIDATION_TYPES.ARRAY,
```
Comment (hetunandu) [inline app/client/src/widgets/TableWidget/index.tsx:63]: Derived values should not need to be validated.

File: app/client/src/workers/validations.ts (hunk 1 above)
Comment (riodeuno) [reply app/client/src/workers/validations.ts:250]: I checked, `isValid` was `false`. `parsed` as you can see is `[]`, only `transformed` was a string like `{{Query1.data}}`. I figured this is the culprit.

Comment (hetunandu) [reply app/client/src/workers/validations.ts:250]: `transformed` will be passed in the `evaluatedValues.tableData` and never `tableData`

File: app/client/src/widgets/TableWidget/index.tsx (hunk 2 above)
Comment (riodeuno) [reply app/client/src/widgets/TableWidget/index.tsx]: I've reverted the code. Despite having a validation on `tableData`. `sanitizedTableData` was being evaluated as the string which was passed for evaluation. validation of `sanitizedTableData` is more expensive than checking for string.

File: app/client/src/workers/validations.ts (hunk 1 above)
Comment (riodeuno) [reply app/client/src/workers/validations.ts:250]: Do you need me to look into why the `sanitizedTableData` is still a string despite the `tableData`'s parsed value is `[]`?

This is the code for the derived property.
Screenshot 2021-03-11 at 6 40 13 PM

Comment (hetunandu) [reply app/client/src/workers/validations.ts:250]: Yea dependencies seem to be correct. Can you ensure the tableData is getting set correctly?

Comment (riodeuno) [reply app/client/src/workers/validations.ts]: @hetunandu As you said, I've verified that the `tableData` is getting the correct value, so I've removed the `transformed` value validation.

Comment (hetunandu) [reply app/client/src/workers/validations.ts]: Okay can you raise this as an issue. Till then we can merge this in to avoid circular dependencies

**Review by hetunandu — APPROVED** [review]

**Pull request merged**